tooling/base1/gaggle/srcdoc.man
NAME

    srcdoc - Generate HTML documentation from source files

SYNOPSIS

    perl srcdoc coversheet

DESCRIPTION

    Srcdoc processes a set of C source files to produce documentation in HTML
    format.  You provide a 'coversheet' file that specifies which source
    files to search.  Srcdoc produces output (a file called html.out) that is
    ready for feeding to the HTML preprocessor htmlpp.  You can fine-tune the
    output by playing with the format file (srcdoc.fmt).

    Srcdoc is not a C parser, so expects some help from the writer.  You must
    use certain modest conventions to tell srcdoc which parts of the source
    code are interesting as documentation.  The current version of srcdoc is
    happiest when looking at packages of functions.

    In general srcdoc assumes that:

        1.  A .h file defines prototypes for one or more .c files.
        2.  All public functions in a .c file have their prototypes in a
            single .h file.
        3.  One .h file corresponds to a section of documentation.
        4.  Each public function in a .c file corresponds to a HTML page
            within that section.

    The coversheet file is a plain ASCII file.  Lines starting with '#' are
    comments.  Blank lines are ignored if possible.  The coversheet contains
    a number of 'sections' that srcdoc looks for.  These are:

    <<files>>
        Specifies the list of source (.h and .c) files to parse.  Srcdoc
        parses these files in the order you list them, so you should arrange
        each .h file followed by the .c files it packages.  The order of .h
        files may be important - you should at least order them so that any
        .h files that use other ones come after them.  This is explained in
        the section 'post-processing' later on.
    <<header>>
        Specifies zero or more lines of htmlpp output that are placed at the
        start of the generated output.  Generally the header contains fairly
        standard htmlpp definitions and possibly an .include command for a
        larger block of descriptive text.

    The order of these blocks is not important.  You can use any case.  Place
    the <<...>> tag by itself on a line.

    This is an example of a coversheet file:

    #    Example of a coversheet file

    <<Files>>
    myfunct.h myfunct.c
    <<Header>>
    .------------------------------------------------------------
    .-  This is generated by srcdoc for input to htmlpp
    .------------------------------------------------------------
    .include prelude.def
    .page myfunct.htm=These Great Functions
    .ignore header
    <H1>$(TITLE)</H1>
    .ignore header
    <H2>Overview</H2>
    <P>These functions provide heuristic argument interpretation.  They
    will do more or less something useful, depending on the supplied
    arguments.</P>
    .ignore header
    <H2>Table of Contents</H2>
    .build toc


    Srcdoc handles .h and .c files differently.  This is how it parses a .h
    file.

    1.  It looks for a line containing "<prolog>".  This marks the start of a
        prolog, or descriptive block.  It parses the prolog until it finds
        "</prolog>".  The prolog consists of a number of clauses indicated by
        a some text followed by ":".  This is pretty formal - srcdoc was
        written to enforce a consistent style.  This is a complete prolog
        block:

        Name:       the name of the file (required)
        Title:      some title text (required)
        Package:    what package (optional)
        Written:    date and author (optional)
        Revised:    date and author (optional)
        Synopsis:   description of this header file (optional)
        Copyright:  some text
        Authors:    authors of package (optional)
        Notes:      descriptive notes (optional).  Put the notes on the line
                    following the Notes: keyword.  The notes are formatted
                    exactly as typed-in, using the HTML <PRE> tag.
        License:    ignored

    Any other clause is an error.  This is an example of a .h file prolog
    block:

    /*  --------------------------------------------------<Prolog>-
        Name:       myfunct.h
        Title:      These great functions
        Written:    91/05/20  Joe Bloggs <joe@bloggs.com>
        Revised:    96/04/10  Joe Bloggs <joe@bloggs.com>
        Synopsis:   These functions are great.  They are so great,
                    in fact, that words fail me.
    -----------------------------------------------------</Prolog>-*/

    2.  Srcdoc looks for function prototypes, #defines, and typedef's.  It
        builds a table for each these sets of definitions.

    3.  You can define several <prolog> blocks in a .h file; only the first
        one should have a Name: clause.  Each prolog block defines a section
        in the .h file, and generates a header.  Srcdoc outputs the function
        prototype, #define, and typedef tables in each section independently.

    4.  Srcdoc ignores functions whose name starts with '_'.

    This is how srcdoc parses a .c file:

    1.  It looks for a line containing "[<]".  This marks the start of a
        function header.  The function header marks a public function, and
        ends with a line containing "[>]".

    2.  The function header consists of a number of clauses indicated by a
        some text followed by ":".

        Function:   the name of the function (required)
        Synopsis:   description of function (optional)
        Examples:   examples of usage (optional).  Put the example text on
                    the line following the Examples: keyword.
        Notes:      descriptive notes (optional).  Put the notes on the line
                    following the Notes: keyword.  The notes are formatted
                    exactly as typed-in, using the HTML <PRE> tag.

    Any other clause is ignored.  This is an example of a function header:

    /*  -------------------------------------------------------------[<]-
        Function: count_sheep
        Synopsis: Takes any number, and counts sheep until sleep() is
                  TRUE.  The sheep argument must be positive and greater
                  than zero.
        Examples:
            count_sheep (100);   - moderate insomnia
            count_sheep (1000);  - advanced insomnia
        -------------------------------------------------------------[>]-*/

    3.  The function header Synopsis: clause can contain various tags that
        format specific blocks.  These are:

        <Table> </Table> - start and end a 2-column table; the first word
        comes in column 1; the remainder of the line in column 2.  Each
        entry on the table must be typed on exactly 1 line.  If you use '_'
        underscores in the first column, these are zapped into spaces.
        <List> </List> - start and end a bulleted list; the whole line is
        taken as a list item.  Do not try to bullet the line yourself.
        If a line starts with '\', this character is removed.

        For example:

        Synopsis: Takes any number, and counts sheep until sleep() is
                  TRUE.  These are some typical sheep values:
        <TABLE>
        1_to_9   Very tired.
        10       Quite tired.
        100      Moderate insomnia.
        1000     Advanced insomnia.
        </TABLE>
        \Note:
        <LIST>
        The sheep value must be a positive integer.
        The sheep value must be greater than zero.
        </LIST>

    4.  Srcdoc skips blank lines, and collects the function prototype.  This
        ends with a '{' in column 1.

    5.  Srcdoc collects the function body, which ends with a '}' in column 1.

    Whenever srcdoc needs to generate some output it refers to the formats
    defined in srcdoc.fmt.  This file describes the purpose of each format;
    you should examine this file if you want to know the specifics of the
    HTML generation done by srcdoc.

    Srcdoc does various replacements within the text that it processes.  This
    affects different types of text differently.  Within the Prolog block,
    srcdoc does these replacements:

    1.  "<name@domain.type>" becomes:
        "<A HREF="mailto:name@domain.type">name@domain.type</A>".
    2.  "<protocol://name>" becomes:
        "<A HREF="protocol://name">protocol://name</A>".
    3.  "<" becomes "&lt;".
    4.  ">" becomes "&gt;".
    5.  "<<Table>>..<</Table>>" becomes a table of entries.
    6.  "<<List>>..<</List>>" becomes a bulleted list.
    7.  "function (" when 'function' is a public function becomes:
        "<A HREF="filename#function">name</A> (".

    In a function header Synposis clause, srcdoc does these same
    replacements.  Within an Examples: clause, srcdoc only replaces < and >.
    In a function body, srcdoc zaps function references.

    Srcdoc uses htmlpp's anchor list (anchor.def) to create links to function
    names (point 7 above).  This means that you must run srcdoc and htmlpp
    several times to be really sure that every reference is correctly built.
    You should probably run srcdoc and htmlpp like this:

    srcdoc coversheet           -- first pass
    htmlpp html.out             -- build anchors
    srcdoc coversheet           -- second pass, uses anchors
    htmlpp html.out             -- build output, uses anchors

    Since htmlpp scans its input twice, this gets a bit slow.  So, buy a
    faster machine.  It is worthwhile.

POST-PROCESSING

    Srcdoc creates two further files: sources.lst and headers.lst.  These
    list the .c files and .h files, one per line.  You can use these as a
    project definition.  For instance, we sometimes build build scripts
    starting from these files.


COPYRIGHT AND LICENSE

    Please see srcdoc source code.



changelog.txt
OpenAMQ change log

AMQP changes for 0.92
    - exchange.declare-ok - add properties result
    - queue.declare-ok - add properties result
    - basic.cancel - explicitly document as assertive
    - exchange.delete - explicitly document as assertive
    - queue.delete - explicitly document as assertive
    - queue.unbind - explicitly document as assertive, queue may be gone
    - basic.ack - make synchronous, reply is ack-ok
    - basic.reject - make synchronous, reply is reject-ok
    - restms class
    - direct class
    - content class properties:
        - remove unused properties
        - add timestamping for latency tracking
        - add/clarify routing and addressing properties
    -> SIG for AMQP/0.9.2

TODO
    - redesign console interface
        - use portals
        - use more restful model
    - option to disable console in amq_server
    - projector exchange, generates test data
        - generate N random messages
        - immediately or with precise delays
        - size, max, min
        - message id template
        - body
    - document iCL classes like icl_longstr, asl_field_table, etc. for
      WireAPI developers
    - generate shortref:
        //  Comment
        type result = method (argument prototype)
    - make PAL arrived processing faster when possible
    - remove all content class variability in asl scripts
        - alias amq_content_basic_t -> amq_content_t
    - show broker HA status via amq_shell
    - show connection direct status via amq_shell
    - make web-based amq_shell tool using http server
    - fix/remove settle time in server (before/after socket? federation?)
    - use portals for exchange implementation interface
    - implement fuzzy text search exchange
        - use Leif's algorithms...
        - example for picking up news items, public news server -> RSS
    - implement rotator exchange
        - show how shared queue can be implemented as exchange
    - implement last-held-values for topic exchange
        - trivial type of exchange
    - tune topic exchange binding etc.
        - save all topic keys at shutdown
        - reload known topic keys at startup
    - solve DoS - create too many queues, exchanges, bindings
        - quotas per connection
        - connections per IP address
    - consider how to implement reliable req-resp and pub-sub
        - use persistent exchange, not queue, to back the data
    - implement persistent queues
        - only for get/ack
        - mainly for RESTful access
        - back the queue using SQLite
        - add SQLite to foreign
    - solve security on exchanges, bindings, queues
        - set permissions on declare
            - world, group, owner
            - exchange publish, bind, delete
            - queue consume, delete
            - defaults can be defined in configuration
            - perhaps using group name in item name
        - need concept of arbitrary user groups
        - document as new security model proposal
    - make federation work with failover
        - already records/replays via peering
        - need to include failover detection & recovery
    - make regression test set for server
        - send to shared queue and read back
        - pubsub over topic exchange
        - direct mode publisher
        - direct mode subscriber
        - returned messages, all modes
        - run peering test case
    - send warnings to amq.notify (check CML spec)
    - win32 support improvements
        - msvs2008 project files from Boom
        - win32 installer
        - win32 service interface
        - win32 front-end (maybe portable GUI? Clutter?)
    - redesign queue->consumer->channel using portals
    - also for exchange->binding->queue?


2010-10-06

* fixed: WireAPI was not clearing messages out of IO thread if client thread was too slow.
* fixed: two access violations in amq_console due to improper multithreading code.
* fixed: OPF config framework did not properly free filename property of object class.

2009-05-23

* fixed: returned messages were not triggering wait end (http://www.openamq.org/issue:58)
* added: no wait in wireapi, timeout = -1 (http://www.openamq.org/issue:65)
* added: implemented acknowledgements on shared queues  (http://www.openamq.org/issue:55)
* added: --no-ack option to keep old behaviour

2009-05-22

* added: in Base2/icl, opaque classes and property export/import
* changed: in OpenAMQ, amq_queue_base.icl class merged into amq_queue_basic.

2009-05-01

* added: in amq_server console, you can see all bindings for a queue (1.4)
* added: in amq_server console, verbose option to see all private queues, must be superuser (1.4)

2009-04-30

* fixed: socket management layer was leaking socket descriptors (1.3, 1.4)
* added: --user and --group options to switch to unprivileged user (1.3, 1.4) (TODO: DOC)
* changed: response to socket exhaustion improved (BASE2-312, 1.3, 1.4)
* added: new option --accept-retry-timeout tunes server response to socket exhaustion (BASE2-312, 1.3, 1.4) (TODO: DOC)
* fixed: WireAPI Direct Mode layer was leaking connections (DAMQ-33, 1.3, 1.4)
* fixed: WireAPI Direct Mode layer could abort at shutdown (AMQ-599, 1.3, 1.4)

2009-04-21

* added: regular expression exchange (http://www.pcre.org, 1.4)
* added: amq.regexp is default regexp exchange
* changed: semi-active failover feature added (http://www.openamq.org/issue:21, 1.3 and 1.4)
* changed: Connection.Start 'id' property renamed to 'connection' for clarity
* added: new Connection.Start property, 'identifier' is server process ID
* fixed: failover detects attempt to connect to self as primary or backup (AMQ-597)
* fixed: federation detects attempt to connect to self (--attach localhost) (AMQ-597)
* fixed: checks channel-nbr from misbehaving clients (http://www.openamq.org/issue:44)

2009-04-20

* changed: added --syslog support (--syslog option) (http://www.openamq.org/issue:51, 1.4)
* changed: some debugging went to daily log, cleaned up (1.4)

tooling/base1/gaggle/srcdoc.d
#----------------------------------------------------------------------------
#   Dialog data and code block - do not modify this code by hand
#   Generated by LIBERO 2.4 on 30 Jul, 2010, 11:06.
#   Schema file used: lrschema.pl
#----------------------------------------------------------------------------

#-- Symbolic constants and event numbers ------------------------------------

$LR_defaults_state = 4;
$LR_STATE_after_init = 0;
$LR_STATE_have_source_file = 1;
$LR_STATE_parsing_include = 2;
$LR_STATE_parsing_code = 3;
$LR_STATE_defaults = 4;

$terminate_event = -1;                  # Halts the dialog
$LR_NULL_EVENT = -2;
$code_event = 0;
$define_event = 1;
$empty_event = 2;
$error_event = 3;
$exception_event = 4;
$header_event = 5;
$include_event = 6;
$no_more_event = 7;
$ok_event = 8;
$prolog_event = 9;
$proto_event = 10;
$typedef_event = 11;

#-- Standard subroutines used by dialog interpreter -------------------------

sub raise_exception {
    local ($event) = @_;
    $the_exception_event = $event;
    $exception_raised = 1;
}

sub LR_initialise {
    $LR_nextst = "";
    $LR_nextst .= "000000001000";
    $LR_nextst .= "301000210000";
    $LR_nextst .= "020002010222";
    $LR_nextst .= "030003010333";
    $LR_nextst .= "000040000000";

    $LR_action = "";
    $LR_action .= "000000020000000001000000";
    $LR_action .= "040005000000030200000000";
    $LR_action .= "000700000010001100060908";
    $LR_action .= "001200000013000500121212";
    $LR_action .= "000000000200000000000000";

    $LR_module = "";
    $LR_module .= "01020304" . "00";
    $LR_module .= "05" . "00";
    $LR_module .= "060708" . "00";
    $LR_module .= "090708" . "00";
    $LR_module .= "04" . "00";
    $LR_module .= "101108" . "00";
    $LR_module .= "1208" . "00";
    $LR_module .= "1308" . "00";
    $LR_module .= "1408" . "00";
    $LR_module .= "1508" . "00";
    $LR_module .= "1004" . "00";
    $LR_module .= "08" . "00";
    $LR_module .= "16171808" . "00";

    $LR_offset [1] = 0;
    $LR_offset [2] = 5;
    $LR_offset [3] = 7;
    $LR_offset [4] = 11;
    $LR_offset [5] = 15;
    $LR_offset [6] = 17;
    $LR_offset [7] = 21;
    $LR_offset [8] = 24;
    $LR_offset [9] = 27;
    $LR_offset [10] = 30;
    $LR_offset [11] = 33;
    $LR_offset [12] = 36;
    $LR_offset [13] = 38;
}

#-- Dialog interpreter starts here ------------------------------------------

$feedback = 0;                          #   Main program feedback
$LR_state = 0;                          #   First state is always 0
&LR_initialise;                         #   Initialise static strings
&initialise_the_program;
while ($the_next_event != $terminate_event)
  {
    $LR_event = $the_next_event;
    $LR_event < 0 || $LR_event >= 12 && die
        "State ", $LR_state,
        " - event ", $LR_event, " is out of range\n";
    $LR_index = ($LR_state * 12 + $LR_event) * 2;
    $LR_index = substr ($LR_action, $LR_index, 2);
    $LR_savest = $LR_state;

    #   If no action for this event, try the defaults state
    if ($LR_index == 0)
      {
        $LR_state = $LR_defaults_state;
        $LR_index = ($LR_state * 12 + $LR_event) * 2;
        $LR_index = substr ($LR_action, $LR_index, 2);
        $LR_index == 0 && die
            "State ", $LR_state,
            " - event ", $LR_event, " is not accepted\n";
      }
    $the_next_event = $LR_NULL_EVENT;
    $the_exception_event = $LR_NULL_EVENT;
    $exception_raised = 0;

    #   Execute module list for state transition, 0 indicates end of list
    $LR_vecptr = $LR_offset [$LR_index];
    $LR_modnbr = substr ($LR_module, $LR_vecptr * 2, 2);
    while ($LR_modnbr > 0 && !$exception_raised)
      {
          {                             #   Execute the module
            $LR_modnbr == 1 && do { &initialise_program_data; last; };
            $LR_modnbr == 2 && do { &load_anchor_definitions; last; };
            $LR_modnbr == 3 && do { &parse_coverpage_file; last; };
            $LR_modnbr == 4 && do { &get_next_source_file; last; };
            $LR_modnbr == 5 && do { &terminate_the_program; last; };
            $LR_modnbr == 6 && do { &may_have_first_include_file; last; };
            $LR_modnbr == 7 && do { &prepare_to_parse_file; last; };
            $LR_modnbr == 8 && do { &look_for_interesting_line; last; };
            $LR_modnbr == 9 && do { &may_have_first_code_file; last; };
            $LR_modnbr == 10 && do { &build_tables_of_definitions; last; };
            $LR_modnbr == 11 && do { &parse_prolog_block; last; };
            $LR_modnbr == 12 && do { &parse_define_command; last; };
            $LR_modnbr == 13 && do { &parse_typedef_command; last; };
            $LR_modnbr == 14 && do { &parse_function_prototype; last; };
            $LR_modnbr == 15 && do { &ignore_function_header; last; };
            $LR_modnbr == 16 && do { &parse_function_header; last; };
            $LR_modnbr == 17 && do { &parse_function_body; last; };
            $LR_modnbr == 18 && do { &build_function_definition; last; };
          }
        $LR_vecptr++;
        $LR_modnbr = substr ($LR_module, $LR_vecptr * 2, 2);
      }

    #   Handle exception if any was raised
    if ($exception_raised)
      {
        $the_exception_event == $LR_NULL_EVENT ||
            ($LR_event = $the_exception_event);

        $the_next_event = $LR_event;
      }
    else
      {
        $LR_index = ($LR_state * 12 + $LR_event) * 1;
        $LR_state = substr ($LR_nextst, $LR_index, 1);
      }
    $LR_state = $LR_savest if ($LR_state == $LR_defaults_state);
    if ($the_next_event == $LR_NULL_EVENT)
      {
        &get_external_event;
        $the_next_event == $LR_NULL_EVENT && die
            "No event set after event ", $LR_event,
            " in state ", $LR_state, "\n";
      }
  }
exit ($feedback);
